### DISCUSSION TITLE
Get items from newly created collection which haven't yet been persisted

### DISCUSSION THREAD

**Discussion by QuestionAndAnswer** [mikro-orm/mikro-orm#3701] [Q&A] [answered]
I'm trying to build runtime mat. view of some data. Let's imaging A is some root aggregate, B has referenced childs, and AView is a mat. view if A with some child props.

```typescript
@Entity()
class B {
  @PrimaryKey()
  id: string;

  @Property()
  name: string;
}

@Entity()
class A {
  @PrimaryKey()
  id: string;
  
  @ManyToMany()
  child = new Collection<B>(this);

  @OneToOne(() => AView, x => x.a, { orphanRemoval: true })
  view: IdentifiedReference<AView>;
}

class AView {
  @PrimaryKey()
  id: string;

  @OneToOne(() => A, x => x.view, { owner: true, orphanRemoval: true })
  a: IdentifiedReference<A>;
  
  @Property()
  cNames: string[];
}

const a = new A();
a.child.set([new B(), new B()]);

const view = new AView();
view.cNames = a.child.getItems().map(x => x.name); // this will fail on getItems() call

a.view = view;

em.persist(a);
em.persist(b);

await em.flush();
```

So, the issue I'm facing, is that trying to read from newly created collection via `getItems()` fails with error that collection wasn't initialized. I can't span this on save - load - save sequence to make A a conventional managed entity. I need to do this in a single transaction. Is there any way to read items from a fresh collection directly, without any intermediate data storage? Seems like it should be possible but I can't find.

Marking collection as populated via `populated()` call is kinda wrong, at least I didn't get from the source the consequences of the call, and there is no documentation on this method.

If there is no way of making `getItems` call, at least `getIdentifiers` maybe? But as I've looked into sources it uses `getItems` internally so I don't think it is the way also.

**Comment by B4nan** [comment, marked as answer]
New collections are initialized, so what you say does not make sense to me.

**Reply by QuestionAndAnswer** [reply]
Hm. Okay, my fault. I tried different approaches and failed to keep the right context.

So, there are couple of methods to load the items.

1. init - does not checks if initialized and will fail for new entities as they are not managed
2. loadItems - requires collections to be fully initialized and also will fail as new collection is only partially initialized (at least that's what I saw in sources)
3. getItems - will return the data if collection is partially initialized.

I just was hoping to write single function which would be able to load data if it is not loaded, but init and loadItems can't work with newly created, already persisted but not flushed entities. So, I was needed to separate mat view update sequence to `loadAndUpdate` and `update`. The first one is loading all the items, the second just relies on getItems calls.

I guess discussion can be closed.

**Reply by B4nan** [reply]
We are talking about new entities, those will never have any items in the database, so not sure what you are expecting here.

It's all the same story, you have a collection, you want to get items from it, so it needs to be initialized. Both populated collections as well as new entities' collections are initialized. You can use `loadItems` to ensure the collection is initialized - this will work for new collections too and is a no-op on them.

> loadItems - requires collections to be fully initialized and also will fail as new collection is only partially initialized (at least that's what I saw in sources)

What makes you think they are "partially initialized"? It's not true. Please back your claims with code snippets at least, so I can understand what you are doing.

Purpose of loadItems is to initialize the collection if it's not already initialized. It should never fail.

> getItems - will return the data if collection is partially initialized.

It will return data always, again not sure what you mean by that claim. It validates if the collection is initialized or not, you can disable that validation via its parameters.

The difference between partially and fully initialized collection is not about the collection at all - it's about the items you put inside - whether they are references (PKs) or fully loaded entities. If your new entity has a partially initialized collection, it only means you've put some entity references inside it (this can happen indirectly, via [propagation](https://mikro-orm.io/docs/propagation)).

**Reply by QuestionAndAnswer** [reply]
>The difference between partially and fully initialized collection is not about the collection at all - it's about the items you put inside - whether they are references (PKs) or fully loaded entities. If your new entity has a partially initialized collection, it only means you've put some entity references inside it (this can happen indirectly, via [propagation](https://mikro-orm.io/docs/propagation)).

Oh. Now that's makes sense. I was looking into https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/entity/Collection.ts#L49 this line and thought that this is always the case when collections is new. But you are right, I've put there references. Thanks for the clarification.
